content/post/azure-monitor-terraform/azure-monitor-terraform.md

```diff
@@ -0,0 +1,240 @@
+canonical: https://medium.com/@dylan.mcginley/azure-monitor-terraform-collecting-custom-log-files-57f74b70567a
+---
+To collect custom log files from a VM (Virtual Machine) managed by Azure, it is recommended to use Azure Monitor and the Azure Monitor Agent (AMA). The Azure Monitor Agent is the successor to the Log Analytics Agent which is due to be deprecated on August 31, 2024. There are currently issues associated with the azurerm provider in relation to Azure Monitor.
+
+I suspect in the future these issues will be addressed with updates but for now this post will serve to provide a workaround along with an explanation as to what these are.
+
+Below are links to some of the capabilities discussed in this article:
+
+* [Azure Log Analytics](https://learn.microsoft.com/en-us/azure/azure-monitor/logs/log-analytics-overview)
+* [Azure Monitor](https://learn.microsoft.com/en-us/azure/azure-monitor/overview)
+* [Azure Monitor Agent](https://learn.microsoft.com/en-us/azure/azure-monitor/agents/agents-overview)
+* [Terraform azurerm provider](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs)
+
+**Azurerm version 3.74.0**
+
+## The issue
+
+Currently the advised method of collecting logs and storing them with Azure Monitor relies on the following configuration
+
+* Setup a Log Analytics Workspace
+* Create a Log Analytics Workspace Table for storing logs
+* Install the correct agents on the VM
+* Setup a Data Collection Rule (DCR)
+* Setup a Data Collection Endpoint
+* Create a Data Collection Association between the Data Collection Rule and the VM
+
+The issue here pertains to the Log Analytics Workspace Table. According to the Azure documentation, it advises that a table be created in the Log Analytics Workspace to store the custom logs. Using the azurerm provider, It is implied that when creating the DCR and setting up the stream declaration that it will create the table in the Log Analytics Workspace, however this does not seem to be the case. Previously developers were able to create the table when setting up the Log Analytics Workspace
+
+```hcl
+resource "azurerm_resource_group" "example" {
+  name     = "example-resources"
+  location = "West Europe"
+}
+
+resource "azurerm_log_analytics_workspace" "example" {
+  name                = "acctest-01"
+  location            = azurerm_resource_group.example.location
+  resource_group_name = azurerm_resource_group.example.name
+  sku                 = "PerGB2018"
+  retention_in_days   = 30
+}
+
+resource "azurerm_log_analytics_workspace_table" "container_logs" {
+  name         = "ContainerLog"
+  workspace_id = azurerm_log_analytics_workspace.example.id
+
+  plan              = "Basic" # or "Analytics"
+  retention_in_days = 7 # per docs, setting to null defaults to workspace default
+}
+```
+
+The resource 'azurerm_log_analytics_workspace_table' has since been deprecated. Currently there is no way to create the Log Analytics Workspace Table using the azurerm provider.
```

Comment:
```suggestion
The resource `azurerm_log_analytics_workspace_table` has since been deprecated. Currently there is no way to create the Log Analytics Workspace Table using the azurerm provider.
```

Please use backtick quotes (`) instead of straight single quotes (')

Reply:
Done